006-dplyr-group-by-ordering.Rmd

```diff
@@ -0,0 +1,161 @@
+#> 2 vctrs          1.3s    1.35s    0.742      308MB    1.11
+```
+
+The reason for the improved performance in the character columns is also the largest potential issue with this proposed change. The current algorithm used by dplyr respects the system locale when ordering character vectors, but the one used by vctrs unconditionally orders character vector groups in the C locale. Ordering in the C locale is much faster, but often produces different results from the system locale (which might be set to English, French, etc.).
+
+This change in group ordering would be seen when a user calls `summarise()` after grouping (or any other function that uses the `group_data()`). If the C locale differs from their system locale, then the grouping columns will be returned in a different order than before. For English users, the main difference with the C locale is in how uppercase vs lowercase letters are ordered:
+
+```{r, eval=FALSE}
+library(dplyr)
+
+df <- tibble(g = c("a", "A", "B", "b", "A", "b"))
+gdf <- group_by(df, g)
+
+# With CRAN dplyr
+summarise(gdf, n = n())
+#> # A tibble: 4 × 2
+#>   g         n
+#>   <chr> <int>
+#> 1 a         1
+#> 2 A         2
+#> 3 b         2
+#> 4 B         1
+
+# With the changes in this tidyup
+summarise(gdf, n = n())
+#> # A tibble: 4 × 2
+#>   g         n
+#>   <chr> <int>
+#> 1 A         2
+#> 2 B         1
+#> 3 a         1
+#> 4 b         2
+```
+
+## Solution
+
+We feel that the potential performance benefits from switching to `vec_locate_sorted_groups()` outweigh the issues that could arise from `group_by()` ordering in the C locale, so the proposed plan is to switch out the current `group_by()` algorithm for the new vctrs implementation, which will begin to order character groups in the C locale unconditionally. We expect all other types to retain the same ordering, having improved performance everywhere, especially when there are many groups.
+
+Documentation for both `summarise()` and `group_by()` would be updated to mention the new behavior with character vectors, and would encourage the usage of `arrange()` after `summarise()` if the returned ordering is important. A pre-release blog post would also mention this change.
+
+## Implementation
+
+-   vctrs PR for exporting `vec_locate_sorted_groups()`
+
+    -   <https://github.com/r-lib/vctrs/pull/1441>
+
+-   dplyr PR for converting to `vec_locate_sorted_groups()`
+
+    -   TODO
+
+## Backwards Compatibility
+
+It is difficult to know how much "in the wild" user code will break as a result of changing from using the system locale to the C locale in `group_by()`. Ideally, users won't be relying on row order for most of their analysis scripts, but we are empathetic to the fact that this is not always the case. Our plan is to run reverse dependency checks and help fix any package code with issues that arise from this, and to provide a pre-release blog post noting this change so that users can update their code as required.
```

Comment:
Do we need to provide a helper to make it easier for affected users to switch back to the old behaviour?

Reply:
We could potentially provide a global option like `dplyr.legacy_group_by_locale = TRUE` that would use the old behavior.

I imagine providing a function like `group_by_legacy()` would be just as annoying as adding an `arrange()` call to actually fix it, so I don't think we'd go that route.

Plus the global option way let's you control group-by calls that you might not control directly.

Ideally when we'd introduce this we'd mention that this is a temporary measure that we'd faze out fairly quickly?

Reply:
I've added the short-term global option, `dplyr.legacy_group_by_locale`, to the proposal